### DISCUSSION TITLE
TextBlock Text is not rendering correctly

### DISCUSSION THREAD

**Discussion by viktorspacil** [AvaloniaUI/Avalonia#15264] [Q&A]
Hello,
does anybody encountered following odd behavior (Apple Silicon M3, MacOS Sonoma 14.4.1, Avalonia UI 11.0.10, FluentIcons 1.1.224, FluentAvaloniaUI 2.0.5) when Text of TextBlock is not rendered correctly (wrong size, foreground or does not shown at all)? When I look onto TextBlocks using Avalonia DevTools then I can see, that properties are set correctly. I am using DynamicResource to set the Text property (to be able to dynamically change the language during application lifetime). Item template looks basically like this:

```axaml
<ControlTemplate>
    <Grid ColumnDefinitions="*,Auto" RowDefinitions="Auto,*">
        <Grid
            Grid.Row="0"
            Grid.Column="0"
            ColumnDefinitions="Auto,*">
            <ic:SymbolIcon
                Grid.Column="0"
                VerticalAlignment="Center"
                FontSize="24"
                Symbol="{TemplateBinding Icon}" />
            <StackPanel VerticalAlignment="Center" Orientation="Vertical" Grid.Column="1" Margin="12 0 0 0">
                <TextBlock Text="{TemplateBinding Title}" 
                           ToolTip.Tip="{TemplateBinding Title}" />
                <TextBlock
                    Foreground="{DynamicResource TextFillColorSecondaryBrush}"
                    Text="{TemplateBinding Description}"
                    TextTrimming="CharacterEllipsis"
                    TextWrapping="NoWrap"
                    ToolTip.Tip="{TemplateBinding Description}"
                    Theme="{StaticResource CaptionTextBlockStyle}" />
            </StackPanel>
        </Grid>
        <ContentPresenter
            Grid.Row="0"
            Grid.Column="1"
            Margin="12,0,0,0"
            Content="{TemplateBinding InnerRightContent}" />
        <ContentPresenter
            Grid.Row="1"
            Grid.Column="0"
            Grid.ColumnSpan="2"
            Margin="0,8,0,0"
            Content="{TemplateBinding Content}"
            IsVisible="{TemplateBinding Content,
                                        Converter={x:Static ObjectConverters.IsNotNull}}" />
    </Grid>
</ControlTemplate>
```

When I do any change on the page (like a user change of the language causing property data change or move to another page and back) then the rendering is fixed. This odd behavior appears randomly when a user is browsing through application pages. It is also interesting that only some of the TextBlocks are affected. Do you have any advice or hint what to check/debug? Thank you.

**Wrong**:
<img width="231" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/6149217/11ae993a-0a44-4903-8f23-02ec885fa1da">

**Correct**:
<img width="233" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/6149217/8260d4b0-b38a-4d15-81e2-0a954fedce70">

**Comment by timunie** [comment]
A controlTemplate is to be used in Styles or ControlThemes only. Can you file a complete minimum sample please?

Note: there's a known issue with dynamic resources in PopUps after switching theme but I doubt it is what we see here.

**Reply by viktorspacil** [reply]
This is the complete theme file:
```axaml
<ResourceDictionary
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:ic="using:FluentIcons.FluentAvalonia"
    xmlns:controls="using:App.View.Controls"
    x:CompileBindings="True">

    <Design.PreviewWith>
        <Border Width="400" Padding="10">
            <controls:SettingsItem
                Title="Test"
                Description="Test Long Description Test Long Description Test Long Description"
                Icon="AirplaneLanding">
                <controls:SettingsItem.InnerRightContent>
                    <Button Content="Test" />
                </controls:SettingsItem.InnerRightContent>
                <ListBox>
                    <ListBoxItem Content="Item #1" />
                    <ListBoxItem Content="Item #2" />
                    <ListBoxItem Content="Item #3" />
                </ListBox>
            </controls:SettingsItem>
        </Border>
    </Design.PreviewWith>

    <ControlTheme x:Key="{x:Type controls:SettingsItem}" TargetType="controls:SettingsItem">
        <Setter Property="Margin" Value="0 8 0 8" />
        <Setter Property="Template">
            <ControlTemplate>
                <Grid ColumnDefinitions="*,Auto" RowDefinitions="Auto,*">
                    <Grid
                        Grid.Row="0"
                        Grid.Column="0"
                        ColumnDefinitions="Auto,*">
                        <ic:SymbolIcon
                            Grid.Column="0"
                            VerticalAlignment="Center"
                            FontSize="24"
                            Symbol="{TemplateBinding Icon}" />
                        <StackPanel VerticalAlignment="Center" Orientation="Vertical" Grid.Column="1" Margin="12 0 0 0">
                            <TextBlock Text="{TemplateBinding Title}" 
                                       ToolTip.Tip="{TemplateBinding Title}" />
                            <TextBlock
                                Foreground="{DynamicResource TextFillColorSecondaryBrush}"
                                Text="{TemplateBinding Description}"
                                TextTrimming="CharacterEllipsis"
                                TextWrapping="NoWrap"
                                ToolTip.Tip="{TemplateBinding Description}"
                                Theme="{StaticResource CaptionTextBlockStyle}" />
                        </StackPanel>
                    </Grid>
                    <ContentPresenter
                        Grid.Row="0"
                        Grid.Column="1"
                        Margin="12,0,0,0"
                        Content="{TemplateBinding InnerRightContent}" />
                    <ContentPresenter
                        Grid.Row="1"
                        Grid.Column="0"
                        Grid.ColumnSpan="2"
                        Margin="0,8,0,0"
                        Content="{TemplateBinding Content}"
                        IsVisible="{TemplateBinding Content,
                                                    Converter={x:Static ObjectConverters.IsNotNull}}" />
                </Grid>
            </ControlTemplate>
        </Setter>
    </ControlTheme>

</ResourceDictionary>
```

I am going to try to create a samle.

**Reply by viktorspacil** [reply]
[render-bug-textblock.zip](https://github.com/AvaloniaUI/Avalonia/files/14904034/render-bug-textblock.zip)

Please see minimum sample in the attachment. I was able to replicate the same "bug" on the windows platform too (Microsoft Windows 10). In the sample you have to keep switching between settings page (gear icon in the top right corner) and main page (back arrow button) (sometimes it takes only few passings, sometimes you have to repeat this for 50 times or more, sometimes helps to change the language, which causes changing of resource files and DynamicResource bindings, or try to change the app theme). If you have no luck then rerun the sample app and try again.
Sometimes you can observe the same odd behavior on the main page when the text "Checking updates..." with loader disappears.

**Reply by timunie** [reply]
sorry, but that is not a minimal sample. It has a lot of code inside where I need to check all the code before running the sample. So please strip down the sample to the issue only.

**Reply by viktorspacil** [reply]
Ok, I will try to strip to code to bare minimum.

**Reply by viktorspacil** [reply]
[render-bug-textblock-2.zip](https://github.com/AvaloniaUI/Avalonia/files/14909875/render-bug-textblock-2.zip)
Hello, here is my second try to provide the minimum sample, this is the bare minimum, because it seems that the bug is harder to replicate if the complexity of app gui is lowered. With this sample I have to repeat the steps almost for 100 times (or more) to replicate the odd behavior. You can see the problem even on the main page (missing text, smaller font, see screenshots below). 
I am able to get the bad rendering within 5 repeats in the original application (MacOS or Windows). 
If it helps I am opened to do videocall because we need to resolve this "bug" otherwise it is going to be reported immediatelly after we release the app to the end users. Thank you.

**Wrong**:
![wrong_textblock](https://github.com/AvaloniaUI/Avalonia/assets/6149217/a1b15685-cba7-4cc7-a8dd-68311fe39669)

**Correct**:
![correct_textblocks](https://github.com/AvaloniaUI/Avalonia/assets/6149217/8e220f0f-2593-4997-a95d-eba8de6999b6)

**Reply by jmacato** [reply]
Hi @viktorspacil 

We only offer extended case support such as video call assistance with our Enterprise Support tier customers as listed here:

https://avaloniaui.net/Support

Thanks!

**Reply by viktorspacil** [reply]
Thank you, if we will not resolve this issue within boundaries of this discussion then we will have to use the help of enterprise support.

**Reply by timunie** [reply]
> because it seems that the bug is harder to replicate if the complexity of app gui is lowered.

Well that seems like we are not able to unblock you from pure community support 😢. I think it is a good idea to reach out to the team to get this sorted out via support contract. Thank you.
